### PR TITLE
Mesh tests

### PR DESCRIPTION
**Pull request by trexfeathers**
### Description

Adds a suite of tests for the `Mesh` class.

---

### Test Style

In agreement with @bjlittle (and with tentative support from @pp-mo), these tests are written in a different style that is unit-esque, but focuses on being easier to write and read. You might call them **"pragmatic tests"** or some-such. The basic idea:

* Write tests that look more similar to how a user/developer would use the functionality in the real world.
* Avoid the difficult process of abstracting every test to definitely only test one thing.
* Consider re-using objects for multiple assertions in one test.
    * May achieve more readable, concise testing.
    * Reduces possible resource-bloat caused by the reduced abstraction.

Cons

* Less granular information on failures.
  So an error takes more investigation.
  But not much more, and errors always need some investigation anyway. 

Pros

* Greatly reduced use of `mock`, which is known to be difficult.
  Plenty of stories of `mock` making testing take longer than the functionality.
* Can write more intuitively - faster and less difficult.
* A single test can achieve the same coverage as many unit tests.
  I.e. more coverage for less thinking/writing.

With this style, I was able to write the `Mesh` tests - ~800 lines, ~80 tests, >100 assertions - pretty much fluidly without having to  stop and work out __how__ to write a test. This definitely hasn't been true in the past when trying to achieve proper unit test abstraction.

The amount of time it takes to achieve good coverage with proper unit tests in my opinion often takes longer than the time they save when investigating errors.

This is of course only worth doing if it doesn't make the tests run excessively longer. For reference: testing the `Mesh` class - a fairly rich class - with reasonably good coverage takes ~100ms locally for me.

If this style causes problems / doesn't solve problems in the months ahead, I'm happy to roll back on this idea. And by all means object if you have other concerns!

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Comment by trexfeathers** [comment]
Something jazzy going on with the tests that I don't _think_ can be related to my work. I'm trying running the whole suite locally to see what's up...

**Review by bjlittle — CHANGES_REQUESTED** [review]
@trexfeathers Great stuff!

Push the outcome of our offline conversation and I'll re-review again :+1:

File: lib/iris/experimental/ugrid.py [hunk 1]
```diff
@@ -1657,12 +1688,16 @@ def filters(
...
-            lambda arg: arg if arg is not None else state, args
-        )
+        true_count = len([arg for arg in args if arg])
+        if true_count > 1:
+            # Standard filter behaviour is 'AND', and coord locations are
+            # mutually exclusive, so multiple True cannot return any results.
+            node = edge = face = False
+        elif true_count == 0:
+            # Treat None as True in this case.
+            node, edge, face = [True if arg is None else arg for arg in args]
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/experimental/ugrid.py]: acdf1888ecab38bb7b00260419f220d572aeaffa

**Review by bjlittle — COMMENTED** [review]
Awesome thanks :+1:

**Pull request merged**